rhine-terminal/Main.hs

```diff
@@ -0,0 +1,137 @@
+import Control.Monad.Trans.Class ()
+
+type App = TerminalConcT IO
+
+-- Clocks
+
+type KeyClock = SelectClock TerminalEventClock Char
+
+keyClock :: KeyClock
+keyClock = SelectClock { mainClock = TerminalEventClock, select = selectChar }
+  where
+    selectChar :: Tag TerminalEventClock -> Maybe Char
+    selectChar = \case
+      Right (KeyEvent (CharKey k) _) -> Just k
+      _ -> Nothing
+
+type BeatClock = Millisecond 1000
+
+-- ClSFs
+
+beat :: ClSF App (LiftClock IO TerminalConcT BeatClock) () ()
+beat = (flip T.cons " > " . (cycle " ." !!) <$> count) >-> displayBeat
+
+displayBeat :: ClSF (TerminalConcT IO) cl Text ()
+displayBeat = arrMCl $ \prompt ->  do
+  Position _ column <- getCursorPosition
+  if column /= 0 then do
+    moveCursorBackward column
+    putText prompt
+    setCursorColumn column
+  else putText prompt
+  flush
+
+key :: ClSF App KeyClock () ()
+key = tagS >-> displayKey
+
+displayKey :: ClSF App cl Char ()
+displayKey = arrMCl $ \k -> do
+  putChar k
+  flush
+
+-- Rhines
+
+type DisplayClock = ParClock App (LiftClock IO TerminalConcT BeatClock) KeyClock
+
+mainRhine ::  Rhine App DisplayClock () ()
+mainRhine = beat @@ liftClock waitClock ||@ terminalConcurrently @|| key @@ keyClock
+
+-- Main
+
+main :: IO ()
+main = do
+  hSetBuffering stdin NoBuffering
+  hSetBuffering stdout NoBuffering
+  withTerminal $ \term -> flowTerminal term mainRhine
+  -- withTerminal $ \term -> flowTerminal term mainRhine'
+
+------------
+--
+-- Alternate
+
+beat' :: ClSF App (LiftClock IO TerminalConcT BeatClock) () Text
+beat' = flip T.cons " > " . (cycle " ." !!) <$> count
+
+key' :: ClSF App KeyClock () Char
+key' = tagS
+
+type DisplayClock' = ParClock App (LiftClock IO TerminalConcT BeatClock) KeyClock
+
+displayClock :: DisplayClock
+displayClock = ParallelClock (liftClock waitClock) keyClock terminalConcurrently
+
+display' :: ClSF App cl (Either Text Char) ()
+display' = arrMCl $ \case
+  Left prompt -> do
+    Position _ column <- getCursorPosition
+    if column /= 0 then do
+      moveCursorBackward column
+      putText prompt
+      setCursorColumn column
+    else putText prompt
+    flush
+  Right k -> do
+    putChar k
+    flush
+
+-- Rhine of sources
+
+sensor' :: Rhine App DisplayClock () (Either Text Char)
+sensor' = beat' @@ liftClock waitClock ++@ terminalConcurrently @++ key' @@ keyClock
+
+-- Rhine of sinks
+actuate' :: Rhine App DisplayClock (Either Text Char) ()
+actuate' = display' @@ displayClock
+
+-- Should it be possible to compose rhines with the same clock values?
+-- We don't have clock equality though...
+
+compRhines :: Monad m => Rhine m cl a b -> Rhine m cl b c -> Rhine m cl a c
+compRhines (Rhine sn1 cl) (Rhine sn2 _) = case (sn1, sn2) of
+  (Synchronous clsf1, Synchronous clsf2) -> Rhine (Synchronous (clsf1 >-> clsf2)) cl
+  _ -> undefined -- dont' understand clock trees enough yet
+
+
+mainRhine' :: Rhine App DisplayClock () ()
+mainRhine' = sensor' `compRhines` actuate'
```

Comment:
Can you instead rebase onto https://github.com/turion/rhine/pull/167 and use the new combinators from there to make something like `sensor' @>-^ display'`?